micrometer-core/src/main/java/io/micrometer/core/instrument/binder/jvm/JvmThreadDeadlockMetrics.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright 2024 VMware, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micrometer.core.instrument.binder.jvm;
+
+import io.micrometer.common.lang.NonNullApi;
+import io.micrometer.common.lang.NonNullFields;
+import io.micrometer.core.instrument.Gauge;
+import io.micrometer.core.instrument.MeterRegistry;
+import io.micrometer.core.instrument.Tag;
+import io.micrometer.core.instrument.binder.BaseUnits;
+import io.micrometer.core.instrument.binder.MeterBinder;
+
+import java.lang.management.ManagementFactory;
+import java.lang.management.ThreadMXBean;
+
+import static java.util.Collections.emptyList;
+
+/**
+ * {@link MeterBinder} for JVM deadlocked threads. These metrics are somewhat expensive to
+ * collect, so they should not be enabled unless necessary. To enable these metrics, add
+ * the following to your application's code ({@code registry} is the meter registry being
+ * used by your application):
+ * <pre>{@code new JvmDeadlockedThreadMetrics().bindTo(registry);}</pre>
+ *
+ * @author Ruth Kurniawati
+ */
+@NonNullApi
+@NonNullFields
+public class JvmThreadDeadlockMetrics implements MeterBinder {
+
+    private final Iterable<Tag> tags;
+
+    public JvmThreadDeadlockMetrics() {
+        this(emptyList());
+    }
+
+    public JvmThreadDeadlockMetrics(Iterable<Tag> tags) {
+        this.tags = tags;
+    }
+
+    @Override
+    public void bindTo(MeterRegistry registry) {
+        ThreadMXBean threadBean = ManagementFactory.getThreadMXBean();
+
+        Gauge.builder("jvm.threads.deadlocked", threadBean, JvmThreadDeadlockMetrics::getDeadlockedThreadCount)
```

Comment:
While the `UnsupportedOperationException` won't be thrown at this time, it doesn't make sense to register a Gauge that will knowingly throw an exception every time the value function is called. We should call ThreadMXBean::findDeadlockedThreads in a try catch block so this Gauge is only registered if the operation is supported. We can log a message instead if it isn't supported.

Reply:
Makes sense - thanks for the suggestion! I've made the changes and added a unit test for it.

Reply:
Thanks again for the quick turnaround. I think ThreadMXBean::findMonitorDeadlockedThreads does not throw `UnsupportedOperationException` which I take to mean it should always be available. Was it intentional that you included both gauges in the try-catch block, meaning both won't be registered? Would it not be useful to users to have the monitor deadlock gauge even if the other is unavailable?

Reply:
I also noticed we can call ThreadMXBean::isSynchronizerUsageSupported rather than using a try-catch block with ThreadMXBean::findDeadlockedThreads.

Reply:
Thanks @shakuzen! I've changed the code to use `ThreadMXBean::isSynchronizerUsageSupported` instead of using a try/catch for `UnsupportedOperationException`. Since`ThreadMXBean::findMonitorDeadlockedThreads` is always available, I've made the registration of `jvm.threads.deadlocked.monitor` unconditional.